### ISSUE TITLE
AbstractSymbolResolutionTest.isJavaVersion9OrAbove() fails for jdkVersion=11

### ISSUE THREAD

**Issue by hecaidi** [javaparser/javaparser#2344]
```
   protected boolean isJavaVersion9OrAbove() {
        String jdkVersion = System.getProperty("java.version");
        return Integer.parseInt(jdkVersion.substring(0, jdkVersion.indexOf('.'))) >= 9;
    }
```
As there is no point when jdkVersion is 11, a StringIndexOutOfBoundsException is thrown

**Comment by hecaidi** [comment]
I am not able to add a label to the issue, so I have added [Bug report] in the title

**Comment by malteskoruppa** [comment]
Thanks! Should be an easy fix.

**Comment by matozoid** [comment]
The story behind it is that a big chunk of the symbol solver test cases run with the compiling JDK on the classpath - which is a big stink of course since the JDK changes from version to version. This way you are required to compile JP with a certain JDK version. That is the underlying problem that needs to be solved, but I bet nobody wants to spend the effort required :-)

**Comment by malteskoruppa** [comment]
Ah yes... it's true that the symbol solver sometimes uses the classloader of the current JDK to do the resolving (that's the `ReflectionTypeSolver`), and thus of course the outcome of a symbol resolution process may depend on the executing JDK.

So for example, if we try to resolve a call `o.isEmpty()` where `o` is an `Optional`, then this will only succeed if we're using a JDK 11 to run JP, because that call is resolved using the internal classloader, but `Optional#isEmpty()` has only been introduced since Java 11 and did not exist in earlier versions.

Hmm... well, I think it would quite a bit of (boring) effort to collect and verify the results of the symbol solving tests when run with each different major JVM version, let alone to maintain that in the long run. :-)

Granted, JP requires at least JVM 8 to compile anyway, so we'd have to maintain that only for JVM versions 8 and above, but that's still 5 different JDK versions right now (8 to 12), with a new version added every 6 months. ;-)

To circumvent this problem, would it be possible to require at least JVM 11 to compile JP in the future? I mean, Java 11 is an LTS release, just as Java 8, and it's been out for some time now... wouldn't it be a good point to switch to Java 11 for JP? That would also allow using some of the cool new Java language features in the JP code itself. Plus, it would solve those symbol resolution problems, since it'll only be possible to compile JP with Java 11 and above anyway. Granted, one could again write symbol resolution tests that work only with Java 12 but not with Java 11, but I don't think there are any tests like that at the moment.

**Comment by matozoid** [comment]
Let's see...
- the thing to fix is to stop the symbol solver tests from using the compiling JDK. We could let Maven download a bunch of JDK jars and use them instead. Still a boring task, but only doing the fun stuff doesn't lead to a good product :-)
- Travis compiles JP with JDK 11. I use 11 too, but have 8 as default and forget to switch all the time.
- Most of the world is on JDK 8. JP must be usable from 8. Compiling can be 11, but the end result has to be Java 8 compatible jars. I personally find it a hassle to manage this :-) It would be really nice to be able to use text blocks for unit tests though...

**Comment by matozoid** [comment]
It was automatically closed, but the issue isn't really really fixed.

**Comment by malteskoruppa** [comment]
> the thing to fix is to stop the symbol solver tests from using the compiling JDK. We could let Maven download a bunch of JDK jars and use them instead. Still a boring task, but only doing the fun stuff doesn't lead to a good product :-)

Interesting! Here's a few thoughts...

Pro:
What I really like about the idea is that it would be possible to selectively adapt the symbol resolving process (for calls into the standard library) to the configured language level for parsing. So when parsing in Java 8 mode, the JDK 8 jar could be used for the resolving process, while in Java 11 mode, the JDK 11 jar could be used, etc. That would be quite nice: respect the language level configured by the user not only for the parsing process, but also for the symbol solving process. It would even allow running JP using a JDK 8, yet still parse Java 12 code (that's possible right now already) _and_ do the resolving according to the model of the standard library defined by Java 12 (that is currently not possible when running JP using JDK 8).

Contra:
It would essentially render the `ReflectionTypeSolver` useless. Instead of the `ReflectionTypeSolver`, one would now use a `JarTypeSolver` which uses some JDK jar to do the resolving. If the `ReflectionTypeSolver` is useless, it wouldn't make sense to keep it around, would it? But somehow it would be sad to discard the `ReflectionTypeSolver` completely, as the logic to use the internal classloader instead of a jar is actually quite elegant. :-) In essence: The whole idea would boil down to remove functionality (the `ReflectionTypeSolver`) and replace it with some already existing functionality (the `JarTypeSolver`) instead...

I am also not quite sure how to _get_ the JDK jars. As far as the compilation process of JP is concerned, sure one could download them using Maven, but the other question is, how to get the right jar if a user wants to just do some symbol solving while using JP? Then the jar has to be available as well. I guess it's out of the question to deliver all JDK jars together with JP (e.g., as resources), as that would make the JP jar size explode. :-) So... specify these JARs as runtime dependencies for JP? But that would also mean that they'll all have to be downloaded when the user packages their own project which depends on JP. Or require users to get the jar they need themselves? But that would be much less user-friendly than the `ReflectionTypeSolver` at the moment, which can simply be instantiated without requiring any additional resources; it would also mean that the language level configured by the user could not be automatically used for the symbol solving process as well (as I dreamed above) because the user would implicitly choose the language level for the symbol solving process by selectinging a particular jar.

Questions, questions... :-)

**Comment by matozoid** [comment]
Answers!
- In my opinion the `ReflectionTypeSolver` is a quick & dirty thing because you're analysing a piece of code, and also the running JDK?! To me it is already recommended behaviour not to use the `ReflectionTypeSolver` when you're serious about your analysis.
- what I'm suggesting is only that we do this for the test suite, here we have the ability to download JDK jars.
- there is an issue somewhere to create a typesolver that is only an index of a certain jar. This way we could supply a Jdk12TypeSolver that does not have any of the Jdk12 code, only the relevant signatures of the classes and methods. When I tried implementing this I got tangled up in a lot of rather clumsy code in the type solver hierarchy and abandoned it again.

**Comment by malteskoruppa** [comment]
> In my opinion the `ReflectionTypeSolver` is a quick & dirty thing because you're analysing a piece of code, and also the running JDK?! To me it is already recommended behaviour not to use the `ReflectionTypeSolver` when you're serious about your analysis.

I guess you're right about that. :-)

When you're serious about your analysis, it is probably best to use a `JarTypeSolver` with a JDK jar of your choice rather than the `ReflectionTypeSolver`.

> what I'm suggesting is only that we do this for the test suite, here we have the ability to download JDK jars.

Ok, but would that not in effect mean: In all tests that currently use the `ReflectionTypeSolver`, use a `JarTypeSolver` instead (with some JDK jar downloaded by Maven)?
And if so, would that not effectively imply that the `JarTypeSolver` gets a lot more tests, while all tests for the `ReflectionTypeSolver` are dropped completely? (because, well, all the `ReflectionTypeSolver` tests would be converted to `JarTypeSolver` tests)
Basically, it sounds like - the `ReflectionTypeSolver` is hard to test for all the known reasons, so we simply don't test it at all (and in exchange test the `JarTypeSolver` a lot more). But that would also mean that JP comes with a component that is not at all covered by the tests, namely, the `ReflectionTypeSolver`.

But perhaps I misunderstand what you are aiming to do? :-)

> there is an issue somewhere to create a typesolver that is only an index of a certain jar. This way we could supply a Jdk12TypeSolver that does not have any of the Jdk12 code, only the relevant signatures of the classes and methods.

Sounds like a really cool improvement of the `JarTypeSolver`! :-)

**Comment by matozoid** [comment]
Now you are right about the `ReflectionTypeSolver` getting a lot less coverage - it sounds like a good time to give it a set of quality unit tests instead of the current "let's see what happens when we chuck 10000 lines of code into the symbol solver" tests. Which were fine to get a good start, but are getting rather unwieldy now :-)

We could write that unit test for a class that has been around forever, and that never had a change. Maybe `Date` since it looks like no amount of deprecation will get rid of that class.

**Comment by malteskoruppa** [comment]
So that would involve two sub-issues then:

1. Change the current `ReflectionTypeSolver` tests to `JarTypeSolver` tests, where the `JarTypeSolver` uses a JDK jar that is downloaded by Maven (may involve adding a build-time dependency for the [maven-download-plugin](https://github.com/maven-download-plugin/maven-download-plugin) or similar).
2. Add some new specific and well-defined `ReflectionTypeSolver` tests that rely on long-standing and stable classes such as `Date`.

Agreed? :-)

**Comment by matozoid** [comment]
Agreed! Create new issues if you want.

I tried having maven download all those test codebases that you see in the symbol solver test project once. That wasn't too hard with the "source plugin". This would need another plugin (since we want jars) but I thought I'd mention it :-)

**Comment by malteskoruppa** [comment]
Ok. For me, just this one issue is fine.

I will attempt to have a look, let's see how it works out. I may not get around to this for another 2 or 3 weeks, though, as I have yet a few other things in the pipeline. :-)

**Comment by matozoid** [comment]
Consider yourself assigned!
